strawberryfields/backends/bosonicbackend/backend.py

```diff
@@ -54,6 +69,207 @@ def __init__(self):
+                # if the op returns a measurement outcome store it in a dictionary
+                val = cmd.op.apply(cmd.reg, self, **kwargs)
+                if val is not None:
+                    for i, r in enumerate(cmd.reg):
+                        if r.ind not in self.ancillae_samples_dict.keys():
+                            self.ancillae_samples_dict[r.ind] = [val[:, i]]
+                        else:
+                            self.ancillae_samples_dict[r.ind].append(val[:, i])
+
+                applied.append(cmd)
+
+            # Rest of operations applied as normal
+            if type(cmd.op) not in nongausspreps + ancilla_gates:
+                try:
+                    # try to apply it to the backend and, if op is a measurement, store outcome in values
+                    val = cmd.op.apply(cmd.reg, self, **kwargs)
+                    if val is not None:
+                        for i, r in enumerate(cmd.reg):
+                            samples_dict[r.ind] = val[:, i]
+
+                            # Internally also store all the measurement outcomes
+                            if r.ind not in all_samples:
+                                all_samples[r.ind] = list()
+                            all_samples[r.ind].append(val[:, i])
+
+                    applied.append(cmd)
+
+                except NotApplicableError:
+                    # command is not applicable to the current backend type
+                    raise NotApplicableError(
+                        "The operation {} cannot be used with the Bosonic Backend.".format(cmd.op)
+                    ) from None
+
+                except NotImplementedError:
+                    # command not directly supported by backend API
+                    raise NotImplementedError(
+                        "The operation {} has not been implemented in the Bosonic Backend for the arguments {}.".format(
+                            cmd.op, kwargs
+                        )
+                    ) from None
+
+        return applied, samples_dict, all_samples
+
+    def init_circuit(self, prog):
+        """Instantiate the circuit and initialize weights, means, and covs
+        depending on the Preparation classes.
+
+        Args:
+            prog (object): sf.Program instance
+
+        Raises:
+            NotImplementedError: if Ket or DensityMatrix preparation used
+        """
+        from strawberryfields.ops import (
+            Bosonic,
+            Catstate,
+            DensityMatrix,
+            Fock,
+            GKP,
+            Ket,
+            _New_modes,
+        )
+
+        nongausspreps = (Bosonic, Catstate, DensityMatrix, Fock, GKP, Ket)
+        nmodes = prog.init_num_subsystems
+        self.begin_circuit(nmodes)
+        # Dummy initial weights, means and covs
+        init_weights, init_means, init_covs = [[0] * nmodes for i in range(3)]
+
+        vac_means = np.zeros((1, 2), dtype=complex)  # .tolist()
```

Comment:
In general the means are complex-valued, so even in this case where they are real-valued we set them to be complex because the could ultimately be multiplied with complex-valued means or covariances in the simulation, and we don't want a type error popping up.